### PR TITLE
signal-desktop: Added update script

### PR DESCRIPTION
**Pull request by Arjun31415**
Signal updates at a very fast pace, so I decided to add an update script.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Arjun31415** [comment]
After running script the diff is - 
```patch
diff --git a/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix b/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
index b877f2fb97db..90cc34c6e921 100644
--- a/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
+++ b/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix
@@ -1,12 +1,12 @@
 { callPackage }: builtins.mapAttrs (pname: attrs: callPackage ./generic.nix (attrs // { inherit pname; })) {
   signal-desktop = {
     dir = "Signal";
-    version = "6.40.0";
-    x64-linux_hash = "sha256-vyXHlycPSyEyv938IKzGM6pdERHHerx2CLY/U+WMrH4=";
+    version = "6.41.0";
+    x64-linux_hash = "sha256-DwlynyMjUVcyM6K1lSUTmG1Vgw3juJFCrVZnJa5qyXg=";
   };
   signal-desktop-beta = {
     dir = "Signal Beta";
-    version = "6.40.0-beta.2";
-    x64-linux_hash = "sha256-pfedkxbZ25DFgz+/N7ZEb9LwKrHuoMM+Zi+Tc21QPsg=";
+    version = "6.42.0-beta.1";
+    x64-linux_hash = "sha256-H+7OvosWQSfEr5nXIba3ArtSgCGkUz4XLG0LkfOcudE=";
   };
 }
```

**Comment by Arjun31415** [comment]
There seems to be only 1 issue. The version for latest release is got from github release page https://github.com/signalapp/Signal-Desktop/releases/ . But the app (.deb) is got from `updates.signal.org`. So it might take some time for the new `.deb` file to be available for download after a new release

**Comment by Arjun31415** [comment]
Requesting review from @bkchr @Mic92

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
   signal-desktop = rec {
     dir = "Signal";
-    version = "6.40.0";
-    version-aarch64 = "6.40.0";
-    url = "https://updates.signal.org/desktop/apt/pool/s/signal-desktop/signal-desktop_${version}_amd64.deb";
-    url-aarch64 = "https://github.com/0mniteck/Signal-Desktop-Mobian/raw/${version-aarch64}/builds/release/signal-desktop_${version-aarch64}_arm64.deb";
-    hash = "sha256-vyXHlycPSyEyv938IKzGM6pdERHHerx2CLY/U+WMrH4=";
-    hash-aarch64 = "sha256-3Pi0c+CGcJR1M4ll51m+B5PmGIcIjjlc0qa9b8rkMeU=";
+    x64-linux_version = "6.40.0";
+    aarch64-osx_version = "6.40.0";
```
Comment (bkchr) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:5]: Why osx? This is also for Linux.

Comment (Arjun31415) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:5]: Sorry, I was confused I'll change it

**Comment by Mic92** [comment]
Thanks for your script, I started adding `passthru.updateScript` to that also ryantm-r can fetch it and noticed that I than need to separate signal-desktop and signal-desktop-beta.
I switched out the custom replace logic with nix-update, which makes it shorter: https://github.com/NixOS/nixpkgs/pull/273147

**Comment by Arjun31415** [comment]
I am unable to leave comment on the changes, so I am writing it here. 

Line 20 should be aarch64 update I believe. 

Also could you let me know what is the use case of `nix update` over just replacing with `sed` so that I can leverage the same in future scripts.

**Comment by Mic92** [comment]
> I am unable to leave comment on the changes, so I am writing it here.
> 
> Line 20 of the script should be aarch64 update I believe.
> 
> Also could you let me know what is the use case of `nix-update` over just replacing with `sed` so that I can leverage the same in future scripts, and how does it work with the bot auto-updating ?

It's also doing hash prefetching in this case. In general it can also do a few other things: https://github.com/Mic92/nix-update/

**Review by Mic92 — APPROVED** [review]
Good to be merged when ofborg evaluation is finished.

File: pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop-aarch64.nix [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{ callPackage }:
+callPackage ./generic.nix { } rec {
+  pname = "signal-desktop";
+  dir = "Signal";
+  version = "6.40.0";
+  url = "https://github.com/0mniteck/Signal-Desktop-Mobian/raw/${version}/builds/release/signal-desktop_${version}_arm64.deb";
+  hash = "sha256-3Pi0c+CGcJR1M4ll51m+B5PmGIcIjjlc0qa9b8rkMeU=";
+}
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop-aarch64.nix:8]: Can we place those into one file for stable and one for Beta?

File: pkgs/applications/networking/instant-messengers/signal-desktop/update.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p bash nix-update curl coreutils
+
+set -ex
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+
+case "$UPDATE_NIX_ATTR_PATH" in
+  signal-desktop)
+    latestTag=$(curl https://api.github.com/repos/signalapp/Signal-Desktop/releases/latest | jq -r ".tag_name")
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/signal-desktop/update.sh:10]: All calls to api.github.com should optional take a GITHUB_TOKEN

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix [hunk 4]
```diff
@@ -1,17 +1,6 @@
...
-  signal-desktop-beta = rec {
-    dir = "Signal Beta";
-    version = "6.40.0-beta.2";
-    hash = "sha256-pfedkxbZ25DFgz+/N7ZEb9LwKrHuoMM+Zi+Tc21QPsg=";
-    url = "https://updates.signal.org/desktop/apt/pool/s/signal-desktop-beta/signal-desktop-beta_${version}_amd64.deb";
-  };
+{ hostPlatform, callPackage }: {
+  signal-desktop = if hostPlatform.system == "aarch64-linux"
+    then callPackage ./signal-desktop-aarch64.nix { }
+    else callPackage ./signal-desktop.nix { };
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:4]:
```suggestion
  signal-desktop = ./signal-desktop.nix { };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop-aarch64.nix (hunk 2 above)
Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/signal-desktop-aarch64.nix:8]: No. nix-update gets confused by this.

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix (hunk 4 above)
Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:4]: Can't do this for the reason above.

File: pkgs/applications/networking/instant-messengers/signal-desktop/update.sh (hunk 3 above)
Comment (Mic92) [reply pkgs/applications/networking/instant-messengers/signal-desktop/update.sh:10]: fixed

**Comment by teutat3s** [comment]
To enable easy backports for signal-desktops frequent updates via labels, this also needs backporting. Let's try this.

**Pull request merged**